### DISCUSSION TITLE
AMD Surface Laptop 4 support (touchscreen, suspend)

### DISCUSSION THREAD

**Discussion by foxharp** [linux-surface/linux-surface#1081] [Q&A]
My wife got a Surface Laptop Go a couple of years ago, and runs Mint on it.  I've been pretty impressed by it, and would get one for myself, except that I really need a backlit keyboard.

So I was just about to buy an SL4 AMD, when it occurred to me I should check on support here.

The Supported Features chart implies that on the AMD SL4, the touchscreen isn't supported and never will be.  True?  I searched these discussions before posting, and it sounds like it might work, but that if you make it work then suspend doesn't.  But I might have misread that.

Anyway:  can someone summarize the state of support (in general, and for the screen and suspend in particular) for the SL4A?

**Comment by qzed** [comment]
I don't really know how well suspend works since I haven't heard any reports lately.

However, touch does not work. This still requires the SPI-HID protocol driver, which we currently don't have for AMD-based devices (or rather any ACPI-based ones that use it). There is a version available for device-tree-based devices, which currently works on the Surface Pro X (we boot that via device trees rather than ACPI). We very briefly tried to make that work on the AMD/ACPI-based devices, but it didn't work out (mostly because it's hard to work on this without having such a device). So until there is someone who owns an AMD-based device, ideally has some knowledge in kernel or lower-level/embedded programming, and is willing to spend a couple hours or days on this, we probably won't be able support it.

**Comment by foxharp** [comment]
Thank you for that summary.  Certainly not having the device would make driver development much harder!

As it happens, I think I can check all of the boxes in your last sentence, except for currently owning the device.  I retired some time ago, but I do have linux kernel/driver and embedded experience.  With a bit of hand-holding on the build and github logistics, I suspect we could figure it out.  My work was mostly on embedded linux, with just a little PC work toward the end.  I do use git, including extensively at work, but I haven't used github in particular in a collaborative fashion.  Again, I'm sure I could figure it out. 

Can kernel builds be done on an 8G machine these days?  Or, perhaps better -- can driver/platform work be done without a full build?

If you have high confidence that the Surface Pro X driver could be made to work, then I'd be willing to get the laptop, knowing that I might end up with a non-working touchscreen in the end.  I've never had a touchscreen on a laptop before, so it wouldn't be the end of the world.

I'm much more concerned about suspend.  Lack of that would be a deal breaker.  (But I don't care much about hibernation -- unless it works better than suspend.  ;-) )

I found this, which implies suspend works, though it's not clear how much power it draws, nor what wakeup states might be working now.  https://github.com/linux-surface/linux-surface/discussions/584#discussioncomment-1432923

And here (https://github.com/linux-surface/linux-surface/discussions/581?sort=new#discussioncomment-1759654 ), DrBrad implies that working suspend depends on nomodeset.  If that's true:  in general, nomodeset is just a boot-time aesthetic thing, right?  Or does it have a larger effect (performance?  function?) than that.    I guess that question goes for "amd_iommu=off iommu=off".  What are the downsides of needing those?

I did read through 458 (https://github.com/linux-surface/linux-surface/issues/458) and was getting excited for a thrilling conclusion, but then it just petered out.  :-)   (Parts of it did bring back some not-so-great memories -- I managed to remember how much I disliked ACPI.)

**Reply by qzed** [reply]
Awesome!

Yes I think an 8G machine should be perfectly fine. The GitHub runners that we're using have 2 cores and 7G ram, so in the worst case you can just limit the number of processes and it will take a bit longer. And also yes, for most of this you can probably get by with building the module out-of-tree and then manually loading it. At least for the driver itself. It could be possible that there are some bugs/issues in the kernel that need to be addressed to make it work.

I am fairly confident that we can make it work. One potential issue is that there are multiple versions of the SPI-HID spec, but drivers for two have been published (one from the Surface Duo Android sources that we use for the Surface Pro X and one newer version that Microsoft attempted to upstream some time ago; but both for device-tree rather than ACPI). Now, it might be possible that there's a third version of that spec, but the two drivers seem similar enough in most parts that I think we could still make it work.

The `nomodeset` option essentially moves video mode setting (resolution etc.) from the kernel back to user-space (which is the way it was handled on older kernels/systems). I believe everything end-user-facing should still work the same.

IOMMU is somewhat important for security: It essentially prevents processes (malware, buggy kernel drivers) from accessing device memory that they shouldn't be accessing. I believe it doesn't have any other impact, so you can probably live reasonably well without it. I'm not sure if `amd_iommu=fullflush` has been tried yet, it seems that this might be another option.

My hope is that the AMD devs fix this at some point... This should be chipset specific and not Surface specific, so other devices should have similar issues (and I really don't know anything about the GPU and IOMMU code, or where to even begin investigating this).

**Reply by StollD** [reply]
> The `nomodeset` option essentially moves video mode setting (resolution etc.) from the kernel back to user-space (which is the way it was handled on older kernels/systems). I believe everything end-user-facing should still work the same.

It also effectively forces software rendering these days, because the X.org userspace GPU drivers are probably terribly outdated and not installed by default on any distribution.

**Reply by qzed** [reply]
Huh, I did not know that... I thought it was just a way of setting resolution/video modes? If this is still an issue it should be brought up upstream then. I think in that case `nomodeset` is not a good workaround.

**Reply by StollD** [reply]
> Huh, I did not know that... I thought it was just a way of setting resolution/video modes? If this is still an issue it should be brought up upstream then. I think in that case `nomodeset` is not a good workaround.

If I understand the linked post correctly, `nomodeset` is only needed to prevent the GPU driver from crashing the kernel before you can apply the actual workaround (disabling the IOMMU).

A similar thing needs to be done when installing a distribution on a machine with certain modern NVIDIA cards. You need to do the install with `nomodeset`, because otherwise nouveau will crash the GPU before the install has finished. Afterwards you can install the binary driver from NVIDIA and remove `nomodeset`.

**Reply by foxharp** [reply]
Aren't nomodeset and iommu=off both cmdline options?     I'm not seeing how the IOMMU gets disabled "later".

**Reply by StollD** [reply]
yeah but sometimes people just write what they did to solve the problem. not what might work best. and `nomodeset` is quicker to type than all the iommu stuff in the grub UEFI environment

**Reply by qzed** [reply]
> Aren't nomodeset and iommu=off both cmdline options? I'm not seeing how the IOMMU gets disabled "later".

Yeah, I'm wondering that as well. According to our wiki entry and [this comment](https://github.com/linux-surface/linux-surface/discussions/581#discussioncomment-1419697), `nomodeset` seems to be an alternative to the IOMMU options. So you should get by with just `amd_iommu=off iommu=off` (as @StollD mentioned) and potentially `acpi_backlight=vendor`. In that case you should get hardware rendering, if I'm not mistaken.

**Reply by davidkisvari** [reply]
@qzed **Nomodeset is required only for installing the OS**, then amd_iommu=off and iommu=off parameters are required for the AMD GPU to work properly. Strangely enough the **Surface kernel is required for the keyboard/touchpad to operate (according to the support matrix these should work oob with recent kernels, but they do not until the Surface kernel is loaded)** So external input is still mandatory for installation. **Suspend works just fine**, I've been using it for almost 2 years now, exclusively with Ubuntu 22.04, and today I switched to Fedora 38, no issues so far.  I don't have embedded programming knowledge and this is my main device but I would gladly sacrifice some of my free time to help the project if it's not beyond my understanding (I do have some programming experience but it's mostly python and bash so not too relevant). I am not a pro by any means but if there is a chance I would gladly give it a go. Just let me know.

**Reply by qzed** [reply]
> Nomodeset is required only for installing the OS, then amd_iommu=off and iommu=off parameters are required for the AMD GPU to work properly.

Assuming kernels for the install medium and the final installation are equal, you should also be able to use `amd_iommu=off` and `iommu=off` directly at install time instead of `nomodeset`.

> Strangely enough the Surface kernel is required for the keyboard/touchpad to operate (according to the support matrix these should work oob with recent kernels, but they do not until the Surface kernel is loaded).

Keyboard/touchpad support also depends on some kernel options being set. If you have any issues with the install media, you might want to check the config file against https://github.com/linux-surface/linux-surface/blob/master/configs/surface-6.2.config and ask the distro's maintainers to enable the Surface Aggregator related ones. Support is present in the upstream sources, it just needs to be enabled/configured.

**Reply by davidkisvari** [reply]
> Assuming kernels for the install medium and the final installation are equal, you should also be able to use `amd_iommu=off` and `iommu=off` directly at install time instead of `nomodeset`.

Probably but it's easier to start the installer in safe graphics mode and set kernel parameters once installed then to start with that (for me at least).

> Keyboard/touchpad support also depends on some kernel options being set. If you have any issues with the install media, you might want to check the config file against https://github.com/linux-surface/linux-surface/blob/master/configs/surface-6.2.config and ask the distro's maintainers to enable the Surface Aggregator related ones. Support is present in the upstream sources, it just needs to be enabled/configured.

Thanks, I didn't know it's something up to the distro being used. This was my experience with both Ubuntu 22.04 (but that came with an older kernel than required if I'm not mistaken) and Fedora 38 which is fairly recent. It doesn't really matter as I'd use the Surface kernel anyway in hopes that touch/pen will work someday.

**Reply by qzed** [reply]
Ah, I think I might have overlooked something. Fedora should have all required config options enabled, but for the AMD devices, support might depend on [this patch](https://github.com/linux-surface/linux-surface/blob/829928f78a68db3ac9cc9dca4b6bd380951046bd/patches/6.2/0013-amd-gpio.patch).

**Comment by foxharp** [comment]
Sigh -- it turns out the vendor I was going to purchase the SL4A from has a very odd system, where a product can be unavailable for both local purchase _and_ via shipping, but still not be "out of stock".  It wasn't until I had it in my cart and hit "checkout" that I was told I couldn't have it.  Their support chatterer (person, not bot) assures me it will be restocked soon.  We'll see.  So rekindling my driver skills must be put on hold.  As will my new laptop.  :-/
